qir/qat/FunctionReplacementPass/FunctionAnnotatorPass.cpp

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "qir/qat/FunctionReplacementPass/FunctionAnnotatorPass.hpp"
+
+#include "qir/qat/FunctionReplacementPass/FunctionReplacementAnalysisPass.hpp"
+#include "qir/qat/Llvm/Llvm.hpp"
+
+#include <fstream>
+#include <iostream>
+
+namespace microsoft::quantum
+{
+
+bool FunctionAnnotatorPass::isRequired()
+{
+    return true;
+}
+
+llvm::PreservedAnalyses FunctionAnnotatorPass::run(llvm::Module& module, llvm::ModuleAnalysisManager& /*mam*/)
+{
+    bool changed     = false;
+    auto annotations = config_.injectedAnnotations();
+
+    // Removing all function call attributes
+    if (config_.shouldRemoveCallAttributes())
```

Comment:
First of all, note that it will have little to no effect on the final IR to remove these attributes as LLVM will add them back in upon looking at the function declaration. As I recall it, the main reason to have this possibility is that replacing a function call where the attributes differ may result in invalid IR (i.e. replacing a tail call with a non-tail call). The work around is to remove all attributes on the call site, make the replacement and then have LLVM add the attributes back onto the call site. Also, note that the default of this option is `false`.

Reply:
We discussed that in more detail offline, and agreed to proceed. We may want to tweak that and be more discerning in the future.